File: scripts/stats.py
#!/usr/bin/env python3
"""Fetch download counts from GitHub releases API and write stats.json."""

import json
import sys
import os
from urllib.request import Request, urlopen

REPO = "lfaoro/ssm"
API_URL = f"https://api.github.com/repos/{REPO}/releases?per_page=100"
OUTPUT = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "data", "stats.json")


def main():
    req = Request(API_URL, headers={"Accept": "application/vnd.github+json",
                                     "User-Agent": "ssm-stats/1.0"})
    with urlopen(req) as resp:
        releases = json.load(resp)

    known_os = {"linux", "darwin", "freebsd", "netbsd", "openbsd", "solaris", "windows"}
    platforms = {}
    total = 0
    for rel in releases:
        for asset in rel.get("assets", []):
            count = asset["download_count"]
            name = asset["name"]
            if name.endswith((".asc", "checksums.txt")):
                continue
            stem = name
            for ext in (".tar.gz", ".tgz", ".zip", ".deb", ".rpm"):
                if stem.endswith(ext):
                    stem = stem[: -len(ext)]
                    break
            key = stem
            for os_name in sorted(known_os, key=len, reverse=True):
                idx = stem.find(os_name + "_")
                if idx >= 0:
                    key = f"{os_name}/{stem[idx + len(os_name) + 1:]}"
                    break
            platforms[key] = platforms.get(key, 0) + count
            total += count

    result = {"total": total, "platforms": dict(sorted(platforms.items(), key=lambda x: -x[1]))}
    os.makedirs(os.path.dirname(OUTPUT), exist_ok=True)
    with open(OUTPUT, "w") as f:
        json.dump(result, f, indent=2)
        f.write("\n")
    print(f"Updated {OUTPUT}: {total} total downloads")


if __name__ == "__main__":
    main()

File: scripts/stats.go
// Copyright (c) 2025 Leonardo Faoro & authors
// SPDX-License-Identifier: MIT

// Fetch download counts from GitHub releases API and write data/stats.json.
//
// Usage: go run scripts/stats.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

const (
	repo    = "lfaoro/ssm"
	apiURL  = "https://api.github.com/repos/" + repo + "/releases?per_page=100"
)

func main() {
	req, err := http.NewRequest(http.MethodGet, apiURL, nil)
	if err != nil {
		fmt.Fprintf(os.Stderr, "stats: %v\n", err)
		os.Exit(1)
	}
	req.Header.Set("Accept", "application/vnd.github+json")
	req.Header.Set("User-Agent", "ssm-stats/1.0")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		fmt.Fprintf(os.Stderr, "stats: %v\n", err)
		os.Exit(1)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		fmt.Fprintf(os.Stderr, "stats: %v\n", err)
		os.Exit(1)
	}

	var releases []struct {
		Assets []struct {
			Name         string `json:"name"`
			DownloadCount int   `json:"download_count"`
		} `json:"assets"`
	}
	if err := json.Unmarshal(body, &releases); err != nil {
		fmt.Fprintf(os.Stderr, "stats: %v\n", err)
		os.Exit(1)
	}

	knownOS := []string{"freebsd", "netbsd", "openbsd", "solaris", "windows", "linux", "darwin"}
	platforms := make(map[string]int)
	var total int

	for _, rel := range releases {
		for _, asset := range rel.Assets {
			name := asset.Name
			if strings.HasSuffix(name, ".asc") || strings.HasSuffix(name, "checksums.txt") {
				continue
			}
			stem := name
			for _, ext := range []string{".tar.gz", ".tgz", ".zip", ".deb", ".rpm"} {
				if strings.HasSuffix(stem, ext) {
					stem = stem[:len(stem)-len(ext)]
					break
				}
			}

			key := stem
			for _, osName := range knownOS {
				idx := strings.Index(stem, osName+"_")
				if idx >= 0 {
					key = osName + "/" + stem[idx+len(osName)+1:]
					break
				}
			}
			platforms[key] += asset.DownloadCount
			total += asset.DownloadCount
		}
	}

	type result struct {
		Total     int            `json:"total"`
		Platforms map[string]int `json:"platforms"`
	}

	sorted := make(map[string]int)
	keys := make([]string, 0, len(platforms))
	for k := range platforms {
		keys = append(keys, k)
	}
	sort.Slice(keys, func(i, j int) bool {
		return platforms[keys[i]] > platforms[keys[j]]
	})
	for _, k := range keys {
		sorted[k] = platforms[k]
	}

	out := result{Total: total, Platforms: sorted}

	outFile := filepath.Join("data", "stats.json")

	f, err := json.MarshalIndent(out, "", "  ")
	if err != nil {
		fmt.Fprintf(os.Stderr, "stats: %v\n", err)
		os.Exit(1)
	}
	f = append(f, '\n')

	if err := os.WriteFile(outFile, f, 0644); err != nil {
		fmt.Fprintf(os.Stderr, "stats: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Updated %s: %d total downloads\n", outFile, total)
}
